File: backend/src/models/Notification.ts
import { Schema, model, Document, Types } from 'mongoose';

export type NotificationType =
  | 'fee_due'
  | 'result_published'
  | 'assignment_graded'
  | 'assignment_created'
  | 'broadcast'
  | 'system';

export interface INotification extends Document {
  orgId: Types.ObjectId;
  branchId: Types.ObjectId;
  recipientId: Types.ObjectId;
  senderId?: Types.ObjectId;
  type: NotificationType;
  title: string;
  message: string;
  link?: string;
  isRead: boolean;
  readAt?: Date;
  metadata?: Record<string, unknown>;
  createdAt: Date;
}

const notificationSchema = new Schema<INotification>(
  {
    orgId: { type: Schema.Types.ObjectId, ref: 'Organization', required: true },
    branchId: { type: Schema.Types.ObjectId, ref: 'Branch', required: true },
    recipientId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    senderId: { type: Schema.Types.ObjectId, ref: 'User' },
    type: {
      type: String,
      enum: ['fee_due', 'result_published', 'assignment_graded', 'assignment_created', 'broadcast', 'system'],
      required: true,
    },
    title: { type: String, required: true },
    message: { type: String, required: true },
    link: String,
    isRead: { type: Boolean, default: false },
    readAt: Date,
    metadata: { type: Schema.Types.Mixed },
  },
  { timestamps: { createdAt: true, updatedAt: false } }
);

notificationSchema.index({ orgId: 1, branchId: 1, recipientId: 1, isRead: 1 });
notificationSchema.index({ orgId: 1, branchId: 1, recipientId: 1, createdAt: -1 });

export const Notification = model<INotification>('Notification', notificationSchema);

File: mobile/lib/models/notification.dart
class AppNotification {
  final String id;
  final String type;
  final String title;
  final String message;
  final bool isRead;
  final DateTime createdAt;
  final String? link;

  const AppNotification({
    required this.id,
    required this.type,
    required this.title,
    required this.message,
    required this.isRead,
    required this.createdAt,
    this.link,
  });

  factory AppNotification.fromJson(Map<String, dynamic> j) => AppNotification(
        id:        j['_id'] as String,
        type:      j['type'] as String,
        title:     j['title'] as String,
        message:   j['message'] as String,
        isRead:    j['isRead'] as bool? ?? false,
        createdAt: DateTime.parse(j['createdAt'] as String),
        link:      j['link'] as String?,
      );
}
